LE audio unicast client callbacks

Thalley opened the pull request:
Update the unicast client to use the newly introduced stream callbacks, instead of relying on the capabilities callbacks. 

The callbacks can also, if wanted, be used by the unicast server, although not necessary.

Thalley committed:
Bluetooth: Audio: Add stream-related callbacks to stream cb struct

Added stream-related callbacks to the stream callback structure.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: endpoint: Update endpoint callbacks to use stream callbacks

Update the callbacks in endpoint.c to use the stream callbacks
instead of the capabilities callbacks.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Refactor bt_audio_stream_config

Refactor the function to take a an audio stream, instead
of returning it. This effects the internals of the stack in the
sense that we no longer use the internally registered capabilities
to retrieve a stream.

This also updates the samples and tests that uses this.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Update unicast client sample and BSIM with stream cbs

Add the newly introduced stream callbacks to the unicast client
sample and the babblesim test cases, and update the implementations
to work with the changes in the stack.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/40036 to build and run tests

asbjornsabo commented on include/bluetooth/audio.h:
```diff
@@ -1172,10 +1172,82 @@ struct bt_audio_broadcast_sink_cb {
+	 */
+	void (*configured)(struct bt_audio_stream *stream);
+
+	/** @brief Stream QoS set callback
```
Wonder whether it would be too general to call it the QoS callback (without the "set").
But thinking about it, the "set" is nice to have, as it gives the same form of the name as the other callbacks.
So - keep as is.

Thalley replied:
We could use the `_cb` suffix you have used for Media Control. It's inconsistent with more callbacks in the BT stack, but it does give some more meaning to the name of the callback.

asbjornsabo replied:
Nah - I was first considering proposing removing the "set", but I think it is better to keep it.
No need for _cb suffix, I think.

asbjornsabo commented on include/bluetooth/audio.h:
```diff
@@ -1172,10 +1172,82 @@ struct bt_audio_broadcast_sink_cb {
 	 *
 	 *  If this callback is provided it will be called whenever the
-	 *  connection completes.
+	 *  isochronous stream is usable.
```
I would rather say "becomes ready to use", or "becomes usable"
As for "whenever", that seems to indicate that is happens many times.  Could that be replace with "when"?

Most importantly: Clarify the difference between this callback and the "started" callback - both are currently saying that they are called when the stream is "usable". 
Or am I just missing the distinction between "stream" and "isochronous stream"?
Should this callback's @brief be with "Isochronous stream connected callback"?  (Ditto below.)

Thalley replied:
Good feedback. 

This callback was added before the `started` callback, and essentially the actual callback that an application should use would be the `started` callback, and not this, as any data would be dismissed until the remote endpoint is in the `streaming` state (which triggers the `started` callback). The ISO stream shall be connected before the remote endpoint can enter this state. 

Can you think of a scenario where the upper layer may want to know whether the ISO is connected on its own? i.e. should we remove the stream connected and disconnected callbacks?

asbjornsabo replied:
Is the upper layer ever doing anything with the ISO?
The abstraction for the upper layer is the (audio) stream, and as long as the ISO is handled invisibly to the upper layers, it should be OK to remove them.

Thalley replied:
It is not. 

The callbacks were added as we didn't have any way earlier to determine whether or not we could actually start sending data, i.e. the client would do `bt_audio_stream_start()` and once that was accepted and completed, it would work, but there was no callback for it. It worked fine for the shell (primarily using debug messages), but not for proper applications. 

However, these may still prove useful for the broadcast sink, as there is no `start` and `stop` callback for that. We could reuse the same callbacks for broadcast. 

What do you think?

asbjornsabo replied:
So - keep the iso stream callbacks and use the for the broadcast sink?
How about rather wrapping them to the same abstraction level (audio stream) as for unicast?

Thalley replied:
We could remove the connected and disconnected ones, and then keep the started and stopped for broadcast sink (possible also for the broadcast source), and then keep the remaining for unicast only.

asbjornsabo replied:
Sounds like a good solution to me.

asbjornsabo commented on subsys/bluetooth/host/audio/endpoint.c:
```diff
@@ -362,31 +366,33 @@ static void ep_config(struct bt_audio_ep *ep, struct net_buf_simple *buf)
 	/* Notify local capability */
-	if (ep->cap && ep->cap->ops && ep->cap->ops->reconfig) {
-		ep->cap->ops->reconfig(ep->stream, ep->cap, ep->stream->codec);
+	if (stream->ops != NULL && stream->ops->configured != NULL) {
```
Line above: Is it still correct to say the we notify capabilites? (Other places as well.)

Thalley replied:
Updated in a new commit.

asbjornsabo commented on include/bluetooth/audio.h:
```diff
@@ -1487,16 +1487,18 @@ int bt_audio_discover(struct bt_conn *conn,
-					       struct bt_audio_capability *cap,
-					       struct bt_codec *codec);
+int bt_audio_stream_config(struct bt_conn *conn,
+			   struct bt_audio_stream *stream,
```
Alignment seems off, but that is probably GitHub.

Thalley replied:
Just Github :) (if you are using Firefox like me, there may be even more weird alignments in Github...)

asbjornsabo commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -550,12 +550,12 @@ static int configure_stream(struct bt_audio_stream **stream)
-					 remote_capabilities[0],
-					 &preset_16_2_1.codec);
-	if (*stream == NULL) {
+	err = bt_audio_stream_config(default_conn, &audio_stream, sinks[0],
```
I guess that `configure_stream()`no longer needs to take a pointer to a pointer to a stream, but could take only a pointer to a stream?

Thalley replied:
Correct, which is also what it does by `&audio_stream` as `audio_stream` is just a statically allocated struct. The function is updated later to use the `stream` from the parameter.

asbjornsabo replied:
I see what happened - I was commenting on a commit in the middle, where the function signature still had the pointer to pointer ...

asbjornsabo reviewed (changes requested):
If I have understood this correctly, I think callback naming and callback documentation should more clearly reflect the difference between audio streams and isochronous streams.

Thalley committed:
Bluetooth: Endpoint: Update comment for calling stream callbacks

Update the comment from "Notify local capability" to
"Notify upper layer" as that is more descriptive now that
we are not using the capabilities.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo reviewed (approved):
My comments have been addressed
The remaining questions about the ISO callbacks can either be changed here (additional commit?) or as anew PR.
